msgraphcore/middleware/retry_middleware.py

```diff
@@ -0,0 +1,206 @@
+        """
+        retry_options = self.get_retry_options()
+        retry_active = True
+        absolute_time_limit = retry_options['timeout']
+        response = None
+        retry_count = 0
+
+        while retry_active:
+            try:
+                start_time = time.time()
+                request.headers.update({'Retry-Attempt': '{}'.format(retry_count)})
+                response = super().send(request, **kwargs)
+                # Check if the request needs to be retried based on the response
+                if self.should_retry(retry_options, response):
+                    retry_active = self.increment_counter(retry_options)
+                    # Get the delay time between retries
+                    sleep_time = self.get_sleep_time(retry_options, retry_count, response)
+                    if retry_active and sleep_time < absolute_time_limit:
+                        time.sleep(sleep_time)
+                        end_time = time.time()
+                        absolute_time_limit -= (end_time - start_time)
+                        retry_count += 1
+                        continue
+                break
+            except Exception as error:
+                raise error
+        return response
+
+    def should_retry(self, retry_options, response):
+        """
+        Determines whether the request should be retried
+        Checks if the request method is in allowed methods
+        Checks if the response status code is in retryable status codes.
+        """
+        if not self._is_method_retryable(retry_options, response.request):
+            return False
+        if not self._is_request_payload_buffered(response):
+            return False
+        return retry_options['total'] and response.status_code in retry_options['retry_codes']
+
+    def _is_method_retryable(self, retry_options, request):
+        """
+        Checks if a given request should be retried upon, depending on
+        whether the HTTP method is in the set of allowed methods
+        """
+        if request.method.upper() not in retry_options['methods']:
+            return False
+        return True
+
+    def _is_request_payload_buffered(self, response):
+        """
+        Checks if the request payload is buffered/rewindable.
+        Payloads with forward only streams will return false and have the responses
+        returned without any retry attempt.
+        """
+        if response.request.method.upper() in frozenset(['HEAD', 'GET', 'DELETE', 'OPTIONS']):
+            return True
+        if response.request.headers['Content-Type'] == "application/octet-stream":
+            return False
+        return True
+
+    def increment_counter(self, retry_options):
+        """
+        Increment the retry counters on every valid retry
+        """
+        if self.retries_exhausted(retry_options):
+            return False
+        retry_options['total'] -= 1
```

Comment:
I think we are managing a similar value in `retry_count` by incrementing with each retry, and `retry_options['total']` by decrementing from the max number of retries on each retry.

Reply:
That is True. `retry_count` is a utility variable as there's a requirement to update the request headers with a `retry_attempt` value that indicates the retry number.